src/crawlee/_utils/context.py

```diff
@@ -0,0 +1,40 @@
+from __future__ import annotations
+
+import asyncio
+from functools import wraps
+from typing import Any, Callable, TypeVar
+
+T = TypeVar('T', bound=Callable[..., Any])
+
+
+def ensure_context(method: T) -> T:
```

Comment:
It would probably be hard to do so without straying even further from how the JS implementation works, but it may be easier to just do something like

```py
class EventManager:
  async def __aenter__(self) -> EventManagerContext:
    pass
  # ... no methods that need to be called on an initialized event manager

class EventManagerContext:
  def emit():
    pass
  # and whatever else needs an initialized context manager
```

... and so on for everything with an initialization step. Honestly, even `BasicCrawler` could benefit from this - the tracking of active state is kinda hairy there as well.

Reply:
This sounds interesting; however, it would mean a more significant change. I'd prefer to approach it as a potential second step following a broader discussion. I will merge this for now, as it works and addresses the desired goals. Feel free to open an issue, and we can continue there, or on Slack.